os/drivers/video/video_null_driver.c

```diff
@@ -0,0 +1,1554 @@
+#include "video_null_driver.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define video_null_givesem(s)		sem_post(s);
+
+#define FILE_FORMAT "/mnt/%dx%d.%s"
+#define FILE_NAMELEN  64
+
+#define MAX_FRAMES 7
+/* Used only in case of discrete */
+#define MAX_INTERVAL 5
+
+#define CHECK_RANGE(value,min,max,step) do { \
+                                          if ((value < min) || \
+                                              (value > max) || \
+                                              ((value - min) % step != 0)) \
+                                            { \
+                                              return -EINVAL;\
+                                            } \
+                                         } while (0)
+#define max_t(a,b) \
+   ({ __typeof__ (a) _a = (a); \
+       __typeof__ (b) _b = (b); \
+     _a > _b ? _a : _b; })
+
+#define min_t(a,b) \
+   ({ __typeof__ (a) _a = (a); \
+       __typeof__ (b) _b = (b); \
+     _a < _b ? _a : _b; })
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static int video_null_open(FAR void *video_priv);
+static int video_null_close(FAR void *video_priv);
+static int video_null_do_halfpush(FAR void *video_priv, bool enable);
+static int video_null_set_buftype(FAR void *video_priv, enum v4l2_buf_type type);
+static int video_null_set_buf(FAR void *video_priv, uint32_t bufaddr, uint32_t bufsize);
+static int video_null_cancel_dma(FAR void *video_priv);
+static int video_null_get_range_of_fmt(FAR void *video_priv, FAR struct v4l2_fmtdesc *format);
+static int video_null_get_range_of_framesize(FAR void *video_priv, FAR struct v4l2_frmsizeenum *frmsize);
+static int video_null_try_format(FAR void *video_priv, FAR struct v4l2_format *format);
+static int video_null_set_format(FAR void *video_priv, FAR struct v4l2_format *format);
+static int video_null_get_range_of_frameinterval(FAR void *video_priv, FAR struct v4l2_frmivalenum *frmival);
+static int video_null_set_frameinterval(FAR void *video_priv, FAR struct v4l2_streamparm *parm);
+static int video_null_get_range_of_ctrlvalue(FAR void *video_priv, FAR struct v4l2_query_ext_ctrl *range);
+static int video_null_get_menu_of_ctrlvalue(FAR void *video_priv, FAR struct v4l2_querymenu *menu);
+static int video_null_get_ctrlvalue(FAR void *video_priv, uint16_t ctrl_class, FAR struct v4l2_ext_control *control);
+static int video_null_set_ctrlvalue(FAR void *video_priv, uint16_t ctrl_class, FAR struct v4l2_ext_control *control);
+static int video_null_refresh(FAR void *video_priv);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
```

Comment:
I think some of these already defined in video.h, refine is not necessary.

Reply:
To make null_video driver non-dependent on video driver, we declared separately